pythonPackagesExtensions: override all Python package sets at once

FRidh opened the pull request:
Python package sets can be overridden by overriding an interpreter
and passing in `packageOverrides = self: super: {...};`. This is fine
in case you need a single interpreter, however, it does not help you
when you want to override all sets.

With this change it is possible to override all sets at once by
appending a list of "extensions" to `pythonPackagesExtensions`.

From reading the implementation you might wonder why a list is used, and
not
`lib.composeExtensions`? The reason is the latter requires knowledge of
the library function. This approach should be easier for most users
as it is similar to how we append to lists of e.g. inputs or patches
when overriding a derivation.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

FRidh commented:
I would still prefer to go via an RFC but I don't see that happening any time soon really.

Infinisil commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9720,6 +9720,10 @@ in
 
+  # Overrides applied to all Python package sets
+  pythonPackagesOverrides = [
+    (self: super: { })
```
This can just be an empty list

FRidh replied:
Correct. I added this here to show what kind of "type" is needed.

Infinisil commented:
On one hand, I'd rather not have yet another override mechanism but make an RFC about it instead. 
On the other hand, we already have so many override mechanisms, one more won't hurt ¯\_(ツ)_/¯

So eh, I'm about 50/50 on doing this.

r-burns commented:
Still important to me. Without something like this, it's impossible to share python package overlays with other people in a configuration-agnostic way.

FRidh commented:
> On one hand, I'd rather not have yet another override mechanism but make an RFC about it instead.
> On the other hand, we already have so many override mechanisms, one more won't hurt ¯_(ツ)_/¯

Not just in mechanisms, but also in names! Overlays, extensions, overrides, ...

I picked this up again because now with flakes being able to override several times becomes even more important.

Infinisil commented:
Coming back to this, because I kind of need something like this at the moment! Because currently even if you try to overlay all packages via `python39Packages = python39Packages.override { overrides = ... }`, you'll still miss `python39.pkgs` and others. So something like this really feels necessary.

Seems like many other people agree too, I'm willing to merge if conflicts are fixed :)

FRidh commented:
Because it came up, I typically use
```
        # Remove when https://github.com/NixOS/nixpkgs/pull/91850 is fixed.
        python3 = let
          self = prev.python3.override {
            inherit self;
            packageOverrides = nixpkgs.lib.composeManyExtensions final.pythonPackagesOverrides;
          };
        in self;

```
in overlays in flakes.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/add-python-package-via-overlay/19783/2

FRidh committed:
pythonPackagesExtensions: override all Python package sets at once

Python package sets can be overridden by overriding an interpreter
and passing in `packageOverrides = self: super: {...};`. This is fine
in case you need a single interpreter, however, it does not help you
when you want to override all sets.

With this change it is possible to override all sets at once by
appending a list of "extensions" to `pythonPackagesExtensions`.

From reading the implementation you might wonder why a list is used, and
not
`lib.composeExtensions`? The reason is the latter requires knowledge of
the library function. This approach should be easier for most users
as it is similar to how we append to lists of e.g. inputs or patches
when overriding a derivation.

FRidh commented:
I've rebased this commit. Also renamed overlays to extensions, to avoid confusion with Nixpkgs-wide overlays.

infinisil commented on pkgs/development/interpreters/python/tests.nix:
```diff
@@ -133,6 +134,19 @@ let
     # in assert myPackages.foobar == myPackages.numpy; myPackages.python.withPackages(ps: with ps; [ foobar ]);
+  } // lib.optionalAttrs (python ? pythonAttr) {
+    # Test applying overrides using pythonPackagesOverlays.
+    # Test is commented out because we should not use pkgs.extend in Nixpkgs
```
This test doesn't look commented out. I think it would be fine to test it like this, `.extend` isn't a huge cost I'm pretty sure

FRidh replied:
Ah yes, I uncommented this to test whether it was still functioning.

In the past usages of `pkgs.extend` were removed.

knedlsepp commented:
Lovely! ❤️

alkasm commented on pkgs/development/interpreters/python/default.nix:
```diff
@@ -19,9 +19,14 @@ with pkgs;
+        # Function that when called
+        # - imports python-packages.nix
+        # - adds spliced package sets to the package set
+        # - applies overrides from `packageOverrides` and `pythonPackagesOverlays`.
```
comment uses the previous name, now pythonPackagesExtensions

alkasm commented on pkgs/development/interpreters/python/tests.nix:
```diff
@@ -133,6 +134,17 @@ let
     #  myPackages = python.pkgs.overrideScope extension;
     # in assert myPackages.foobar == myPackages.numpy; myPackages.python.withPackages(ps: with ps; [ foobar ]);
+  } // lib.optionalAttrs (python ? pythonAttr) {
+    # Test applying overrides using pythonPackagesOverlays.
```
same
